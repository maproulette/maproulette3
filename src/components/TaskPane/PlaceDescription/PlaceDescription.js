import React, { Component } from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import _get from 'lodash/get'
import _isObject from 'lodash/isObject'
import _isString from 'lodash/isString'

export default class PlaceDescription extends Component {
  render() {
    const addr = _get(this.props, 'place.address')
    const scaleDescriptions = []

    if (_isObject(addr)) {
      if (_isString(addr.city)) {
        scaleDescriptions.push(addr.city)
      }
      else if (_isString(addr.town)) {
        scaleDescriptions.push(addr.town)
      }
      else if (_isString(addr.hamlet)) {
        scaleDescriptions.push(addr.hamlet)
      }
      else if (_isString(addr.village)) {
        scaleDescriptions.push(addr.village)
      }

      if (_isString(addr.county)) {
        scaleDescriptions.push(
          addr.county.toLowerCase().indexOf('county') === -1 ?
          `${addr.county} County` :
          addr.county
        )
      }

      if (_isString(addr.state)) {
        scaleDescriptions.push(addr.state)
      }

      if (_isString(addr.country)) {
        scaleDescriptions.push(addr.country)
      }
      else if (_isString(addr.continent)) {
        scaleDescriptions.push(addr.continent)
      }
    }

    return (
      <div className={classNames('place-description', this.props.className)}>
        {scaleDescriptions.length === 0 ? null : scaleDescriptions.join(', ')}
      </div>
    )
  }
}

PlaceDescription.propTypes = {
  place: PropTypes.object,
}
