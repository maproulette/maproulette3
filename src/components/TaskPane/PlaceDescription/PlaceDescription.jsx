import classNames from "classnames";
import _isObject from "lodash/isObject";
import _isString from "lodash/isString";
import { Component } from "react";

export default class PlaceDescription extends Component {
  render() {
    const addr = this.props.address;
    const scaleDescriptions = [];

    if (_isObject(addr)) {
      if (_isString(addr.city)) {
        scaleDescriptions.push(addr.city);
      } else if (_isString(addr.town)) {
        scaleDescriptions.push(addr.town);
      } else if (_isString(addr.hamlet)) {
        scaleDescriptions.push(addr.hamlet);
      } else if (_isString(addr.village)) {
        scaleDescriptions.push(addr.village);
      }

      if (_isString(addr.county)) {
        scaleDescriptions.push(addr.county);
      }

      if (_isString(addr.state)) {
        scaleDescriptions.push(addr.state);
      }

      if (_isString(addr.country)) {
        scaleDescriptions.push(addr.country);
      } else if (_isString(addr.continent)) {
        scaleDescriptions.push(addr.continent);
      }
    }

    return (
      <div
        data-testid="place-description"
        className={classNames("place-description", this.props.className)}
      >
        {scaleDescriptions.length === 0 ? null : scaleDescriptions.join(", ")}
      </div>
    );
  }
}
