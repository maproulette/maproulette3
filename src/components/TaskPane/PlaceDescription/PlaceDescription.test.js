import React from 'react'
import PlaceDescription from './PlaceDescription'
import _cloneDeep from 'lodash/cloneDeep'

const propsFixture = {
  place: {
    address: {
      city: "My City",
      county: "Orange",
      state: "Washington",
      country: "USA"
    }
  },
}

let basicProps = null

beforeEach(() => {
  basicProps = _cloneDeep(propsFixture)
})

test("renders with props as expected", () => {
  const wrapper = shallow(
    <PlaceDescription {...basicProps} />
  )

  expect(wrapper.find('.place-description').getElement().props.children).toBe(
    'My City, Orange County, Washington, USA'
  )

  expect(wrapper).toMatchSnapshot()
})

test("does not explode with null place", () => {
  const wrapper = shallow(
    <PlaceDescription />
  )

  expect(wrapper).toMatchSnapshot()
})

test("renders with props className in encapsulating div", () => {
  const wrapper = shallow(
    <PlaceDescription {...basicProps} className="my-test"/>
  )

  expect(wrapper.find('.my-test').exists()).toBe(true)
  expect(wrapper).toMatchSnapshot()
})

test("if county is in description, does not render word 'county'", () => {
  basicProps.place.address.county = 'Long Beach County'
  const wrapper = shallow(
    <PlaceDescription {...basicProps} />
  )

  expect(wrapper.find('.place-description').getElement().props.children).toBe(
    'My City, Long Beach County, Washington, USA'
  )
  expect(wrapper).toMatchSnapshot()
})

test("renders town if no city", () => {
  basicProps.place.address.city = null
  basicProps.place.address.town = "Our Town"

  const wrapper = shallow(
    <PlaceDescription {...basicProps} />
  )

  expect(wrapper.find('.place-description').getElement().props.children).toBe(
    'Our Town, Orange County, Washington, USA'
  )
  expect(wrapper).toMatchSnapshot()
})

test("renders hamlet if no town and no city", () => {
  basicProps.place.address.city = null
  basicProps.place.address.hamlet = "The Shire"

  const wrapper = shallow(
    <PlaceDescription {...basicProps} />
  )

  expect(wrapper.find('.place-description').getElement().props.children).toBe(
    'The Shire, Orange County, Washington, USA'
  )
  expect(wrapper).toMatchSnapshot()
})

test("renders village if no hamlet, no town and no city", () => {
  basicProps.place.address.city = null
  basicProps.place.address.village = "Village Square"

  const wrapper = shallow(
    <PlaceDescription {...basicProps} />
  )

  expect(wrapper.find('.place-description').getElement().props.children).toBe(
    'Village Square, Orange County, Washington, USA'
  )
  expect(wrapper).toMatchSnapshot()
})

test("renders continent if no country", () => {
  basicProps.place.address.country = null
  basicProps.place.address.continent = "Africa"

  const wrapper = shallow(
    <PlaceDescription {...basicProps} />
  )

  expect(wrapper.find('.place-description').getElement().props.children).toBe(
    'My City, Orange County, Washington, Africa'
  )
  expect(wrapper).toMatchSnapshot()
})
