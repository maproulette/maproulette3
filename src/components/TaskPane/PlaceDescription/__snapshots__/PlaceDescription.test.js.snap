// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does not explode with null place 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PlaceDescription />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": null,
      "className": "place-description",
    },
    "ref": null,
    "rendered": null,
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": null,
        "className": "place-description",
      },
      "ref": null,
      "rendered": null,
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`if county is in description, does not render word 'county' 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PlaceDescription
    place={
      Object {
        "address": Object {
          "city": "My City",
          "country": "USA",
          "county": "Long Beach County",
          "state": "Washington",
        },
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": "My City, Long Beach County, Washington, USA",
      "className": "place-description",
    },
    "ref": null,
    "rendered": "My City, Long Beach County, Washington, USA",
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "My City, Long Beach County, Washington, USA",
        "className": "place-description",
      },
      "ref": null,
      "rendered": "My City, Long Beach County, Washington, USA",
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`renders continent if no country 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PlaceDescription
    place={
      Object {
        "address": Object {
          "city": "My City",
          "continent": "Africa",
          "country": null,
          "county": "Orange",
          "state": "Washington",
        },
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": "My City, Orange County, Washington, Africa",
      "className": "place-description",
    },
    "ref": null,
    "rendered": "My City, Orange County, Washington, Africa",
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "My City, Orange County, Washington, Africa",
        "className": "place-description",
      },
      "ref": null,
      "rendered": "My City, Orange County, Washington, Africa",
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`renders hamlet if no town and no city 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PlaceDescription
    place={
      Object {
        "address": Object {
          "city": null,
          "country": "USA",
          "county": "Orange",
          "hamlet": "The Shire",
          "state": "Washington",
        },
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": "The Shire, Orange County, Washington, USA",
      "className": "place-description",
    },
    "ref": null,
    "rendered": "The Shire, Orange County, Washington, USA",
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "The Shire, Orange County, Washington, USA",
        "className": "place-description",
      },
      "ref": null,
      "rendered": "The Shire, Orange County, Washington, USA",
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`renders town if no city 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PlaceDescription
    place={
      Object {
        "address": Object {
          "city": null,
          "country": "USA",
          "county": "Orange",
          "state": "Washington",
          "town": "Our Town",
        },
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": "Our Town, Orange County, Washington, USA",
      "className": "place-description",
    },
    "ref": null,
    "rendered": "Our Town, Orange County, Washington, USA",
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "Our Town, Orange County, Washington, USA",
        "className": "place-description",
      },
      "ref": null,
      "rendered": "Our Town, Orange County, Washington, USA",
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`renders village if no hamlet, no town and no city 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PlaceDescription
    place={
      Object {
        "address": Object {
          "city": null,
          "country": "USA",
          "county": "Orange",
          "state": "Washington",
          "village": "Village Square",
        },
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": "Village Square, Orange County, Washington, USA",
      "className": "place-description",
    },
    "ref": null,
    "rendered": "Village Square, Orange County, Washington, USA",
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "Village Square, Orange County, Washington, USA",
        "className": "place-description",
      },
      "ref": null,
      "rendered": "Village Square, Orange County, Washington, USA",
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`renders with props as expected 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PlaceDescription
    place={
      Object {
        "address": Object {
          "city": "My City",
          "country": "USA",
          "county": "Orange",
          "state": "Washington",
        },
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": "My City, Orange County, Washington, USA",
      "className": "place-description",
    },
    "ref": null,
    "rendered": "My City, Orange County, Washington, USA",
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "My City, Orange County, Washington, USA",
        "className": "place-description",
      },
      "ref": null,
      "rendered": "My City, Orange County, Washington, USA",
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`renders with props className in encapsulating div 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PlaceDescription
    className="my-test"
    place={
      Object {
        "address": Object {
          "city": "My City",
          "country": "USA",
          "county": "Orange",
          "state": "Washington",
        },
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": "My City, Orange County, Washington, USA",
      "className": "place-description my-test",
    },
    "ref": null,
    "rendered": "My City, Orange County, Washington, USA",
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "My City, Orange County, Washington, USA",
        "className": "place-description my-test",
      },
      "ref": null,
      "rendered": "My City, Orange County, Washington, USA",
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
